### PR TITLE
Support for normal/pretrained embeddings; updated requirements.txt

### PR DESCRIPTION
**Pull request by VIGS25**

### CODE REVIEW

File: codenames/supervised/embeddings.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import torch
+import torch.nn as nn
+from torchnlp.word_to_vector import GloVe
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+
+def load_word2vec(path_to_file, freeze=False, binary=True):
+    """Loads pretrained word2vec embeddings"""
```
Comment (oskopek) [inline codenames/supervised/embeddings.py]:
```suggestion
    """Loads pretrained word2vec embeddings."""
```

File: codenames/supervised/embeddings.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    model = gensim.models.KeyedVectors.load_word2vec_format(path_to_file, binary=binary)
+    embedding_vectors = torch.tensor(model.vectors, dtype=torch.double, device=device)
+    word_to_index = {word: vector.index for word, vector in model.vocab.items()}
+    embedding = nn.Embedding.from_pretrained(embedding_vectors, freeze=freeze)
+    return embedding, word_to_index
+
+
+def load_glove(path_to_file, freeze=False, binary=True):
+    """Loads pretrained GloVe embeddings"""
```
Comment (oskopek) [inline codenames/supervised/embeddings.py]:
```suggestion
    """Loads pretrained GloVe embeddings."""
```

File: codenames/supervised/embeddings.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    model_dict = GloVe(name='840B', dim=300)
+    embedding_vectors = list(model_dict.values())
+    embedding_vectors = torch.tensor(embedding_vectors, dtype=torch.double, device=device)
+    word_to_index = {word: i for i, word in enumerate(list(model_dict.keys()))}
```
Comment (oskopek) [inline codenames/supervised/embeddings.py]:
```suggestion
    word_to_index = {word: i for i, word in enumerate(model_dict.keys())}
```

File: codenames/supervised/embeddings.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    embedding = nn.Embedding.from_pretrained(embedding_vectors, freeze=freeze)
+    return embedding, word_to_index
+
+
+def load_pretrained(path_to_file, embed_type='word2vec', freeze=False, binary=True):
+    """Loads pretrained embeddings based on type"""
```
Comment (oskopek) [inline codenames/supervised/embeddings.py]:
```suggestion
    """Loads pretrained embeddings based on type."""
```

File: codenames/supervised/embeddings.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    if not os.path.exists(path_to_file):
+        raise FileNotFoundError("File does not exist. Please download embeddings and try again.")
+
+    load_func = globals()['load_' + embed_type]
+    return load_func(path_to_file, freeze=freeze, binary=binary)
+
+
+def setup_embeddings(embedding_args):
+    """Loads pretrained embeddings if present, else initializes random embeddings"""
```
Comment (oskopek) [inline codenames/supervised/embeddings.py]:
```suggestion
    """Loads pretrained embeddings if present, else initializes random embeddings."""
```

File: codenames/supervised/embeddings.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    path_to_file = embedding_args.get('path_to_file', None)
+    freeze = embedding_args.get('freeze', False)
+
+    if path_to_file is not None:
+        embedding_type = embedding_args.get('embed_type')
+        return load_pretrained(path_to_file, embed_type=embedding_type, freeze=freeze, binary=True)
+    else:
+        num_embeddings = embedding_args['num_embeddings']
+        embedding_dim = embedding_args['embedding_dim']
+        embedding = nn.Embedding(num_embeddings, embedding_dim)
```
Comment (oskopek) [inline codenames/supervised/embeddings.py]:
```suggestion
        embedding = nn.Embedding(embedding_args['num_embeddings'], embedding_args['embedding_dim'])
```

File: codenames/supervised/network.py [hunk 7]
```diff
@@ -9,37 +9,33 @@
...
         super(EncoderRNN, self).__init__()
         self.hidden_size = hidden_size
-
-        self.embedding = nn.Embedding(input_size, hidden_size)
         self.gru = nn.GRU(hidden_size, hidden_size)
 
-    def forward(self, input, hidden):
-        embedded = self.embedding(input).view(1, 1, -1)
-        output = embedded
+    def forward(self, inputs, hidden):
```
Comment (oskopek) [inline codenames/supervised/network.py]: I would rename the variables here and below do something like `embedding_size` and `embedded`/`embedded_input` and get rid of the comments about inputs being embedded.

File: codenames/supervised/network.py [hunk 8]
```diff
@@ -9,37 +9,33 @@
...
+        # Inputs are already embedded tokens
+        output = inputs
```
Comment (oskopek) [inline codenames/supervised/network.py]: Also, this line weird. Just put it in the GRU directly.

File: codenames/supervised/network.py [hunk 9]
```diff
@@ -9,37 +9,33 @@
...
 
     def __init__(self, hidden_size, clue_vocab_size):
-        super(ClueClassificator, self).__init__()
+        super(ClueClassifier, self).__init__()
 
         self.linear = nn.Linear(hidden_size, clue_vocab_size)
 
-    def forward(self, input, hidden):
-        output = self.linear(input)
+    def forward(self, inputs, hidden):
```
Comment (oskopek) [inline codenames/supervised/network.py]: Same here, embeddedsomething.

File: codenames/supervised/network.py [hunk 10]
```diff
@@ -52,16 +48,15 @@ def __init__(self, hidden_size, output_size, dropout_p=0.1, max_length=ATTN_MAX_
...
         self.attn = nn.Linear(self.hidden_size * 2, self.max_length)
         self.attn_combine = nn.Linear(self.hidden_size * 2, self.hidden_size)
         self.dropout = nn.Dropout(self.dropout_p)
         self.gru = nn.GRU(self.hidden_size, self.hidden_size)
         self.out = nn.Linear(self.hidden_size, self.output_size)
 
-    def forward(self, input, hidden, encoder_outputs):
-        embedded = self.embedding(input).view(1, 1, -1)
-        embedded = self.dropout(embedded)
+    def forward(self, inputs, hidden, encoder_outputs):
```
Comment (oskopek) [inline codenames/supervised/network.py]: Same here.

File: codenames/supervised/train.py [hunk 11]
```diff
@@ -162,9 +167,9 @@ def main():
 
     encoder = network.EncoderRNN(len(board_vocab), hidden_size).to(device)
     clue_scaler = nn.Linear(len(clue_vocab), hidden_size).to(device)
-    clue_predictor = network.ClueClassificator(hidden_size, len(clue_vocab)).to(device)
+    clue_predictor = network.ClueClassifier(hidden_size, len(clue_vocab)).to(device)
     attn_decoder = network.AttnDecoderRNN(hidden_size * 2, len(board_vocab), dropout_p=0.1).to(device)
-    player_trainer = PlayerTrainer(hidden_size, len(board_vocab), len(clue_vocab)).to(device)
+    player_trainer = PlayerTrainer(hidden_size, len(board_vocab), len(clue_vocab))
```
Comment (oskopek) [inline codenames/supervised/train.py]: Great catch!

File: codenames/supervised/embeddings.py (hunk 1 above)
Comment (VIGS25) [reply codenames/supervised/embeddings.py]: Done

File: codenames/supervised/embeddings.py (hunk 3 above)
Comment (VIGS25) [reply codenames/supervised/embeddings.py]: Need to test load_glove and see which one works.

File: codenames/supervised/embeddings.py (hunk 4 above)
Comment (VIGS25) [reply codenames/supervised/embeddings.py]: Done

File: codenames/supervised/network.py (hunk 7 above)
Comment (VIGS25) [reply codenames/supervised/network.py]: Done

File: codenames/supervised/network.py (hunk 8 above)
Comment (VIGS25) [reply codenames/supervised/network.py]: Done

File: codenames/supervised/network.py (hunk 9 above)
Comment (VIGS25) [reply codenames/supervised/network.py]: Done

File: codenames/supervised/network.py (hunk 10 above)
Comment (VIGS25) [reply codenames/supervised/network.py]: Done

File: codenames/supervised/embeddings.py (hunk 5 above)
Comment (VIGS25) [reply codenames/supervised/embeddings.py]: Done

File: codenames/supervised/embeddings.py (hunk 2 above)
Comment (VIGS25) [reply codenames/supervised/embeddings.py]: Done

File: codenames/supervised/embeddings.py (hunk 3 above)
Comment (oskopek) [reply codenames/supervised/embeddings.py]: Honestly both should work. My suggestion should do potentially less computations.

**Comment by jendelel** [comment]
LGTM

File: codenames/supervised/train.py [hunk 12]
```diff
@@ -153,6 +156,8 @@ def main():
     parser = argparse.ArgumentParser()
     parser.add_argument("--board_words", help="Path to file with board words. Each word on a separate line.")
     parser.add_argument("--clue_words", help="Path to file with clue words. Each word on a separate line.")
+    parser.add_argument("--embedding_path", default='', help='Path to pre-trained embeddings')
+    parser.add_argument("")
```
Comment (oskopek) [inline codenames/supervised/train.py]: Hm?

Comment (VIGS25) [reply codenames/supervised/train.py]: Will deal with this part later

File: codenames/supervised/embeddings.py (hunk 3 above)
Comment (VIGS25) [reply codenames/supervised/embeddings.py]: Done.

File: codenames/supervised/train.py (hunk 12 above)
Comment (oskopek) [reply codenames/supervised/train.py]: Do you wanna merge now?

Comment (VIGS25) [reply codenames/supervised/train.py]: Yup

**Pull request merged**